### ISSUE TITLE
Support for net6.0

### ISSUE THREAD

**Issue by pieterh** [samuelventura/SharpModbus#8]
Hi,
Would be great if SharpModbus and SharpSerial would be available for .NET 6.0. I have created a small PR for both of the packages. With these changes they will support 4 and 6.0 target. thanks in advance

**Comment by pieterh** [comment]
@samuelventura any chance you will create a new package version that includes the support for net6.0?

**Comment by samuelventura** [comment]
Yes. Give me a couple days. Thank you for your PR.

**Comment by samuelventura** [comment]
Unfortunatelly, SharpModbus depends on SharpSerial which has EXE OutputType and net6.0 won't include the exe and when manually added MSBuild complains with:

```
MSBUILD : error : This is an unhandled exception in MSBuild -- PLEASE UPVOTE AN EXISTING ISSUE OR FILE A NEW ONE AT https://aka.ms/msbuild/unhandled. [C:\Users\samuel\Desktop\SharpModbus\SharpModbus\SharpModbu
s.csproj]
MSBUILD : error :     Microsoft.Build.Framework.InternalErrorException: MSB0001: Internal MSBuild Error: Unexpected CopyLocal flag. [C:\Users\samuel\Desktop\SharpModbus\SharpModbus\SharpModbus.csproj]
MSBUILD : error :    at Microsoft.Build.Tasks.CopyLocalStateUtility.IsCopyLocal(CopyLocalState state) [C:\Users\samuel\Desktop\SharpModbus\SharpModbus\SharpModbus.csproj]
MSBUILD : error :    at Microsoft.Build.Tasks.ResolveAssemblyReference.LogResults(ReferenceTable dependencyTable, List`1 idealAssemblyRemappings, List`1 idealAssemblyRemappingsIdentities, List`1 generalResolut 
ionExceptions) [C:\Users\samuel\Desktop\SharpModbus\SharpModbus\SharpModbus.csproj]
M
...
```

There is some work in progress on that [issue](https://github.com/dotnet/msbuild/issues/4072) but I am not sure if it would cover our need which is to include exe along dll files for net6.0 targets.

Published SharpSerial v1.0.6 and v1.0.7 were unsuccessful tryouts.

**Comment by pieterh** [comment]
Hi, 
It is a bit late here in the Netherlands but couldn't resist a quick look at the problem. I can't recall I had an issue building those two projects a while ago. I'm using a Mac where the behaviour could be different. I will have a proper look and try on both Windows and Mac. 
I did see the following, which might be wat is needed here. Will let you know tomorrow.
https://learn.microsoft.com/en-us/dotnet/core/deploying/single-file/overview?tabs=cli

**Comment by samuelventura** [comment]
Can you tell me about your use case for this library? I not sure if you have found the isolated serial valuable. It is because of that feature that sharpserial needs to be an exe and causes this trouble.

**Comment by pieterh** [comment]
I'm using it in the communication with my chargepoint for an electric vehicle. In my case it is an "Alfen Eve Single Pro-line". The communication for the Alfen NG9xx platform is using modbus. Using modbus I'm able to determine the charging state, control the amount of current that is available for charging and pause/resume charging. With this it is possible to charge using the lowest cost energy and making sure that I'm not overloading the main switchboard.
For my use case there is no need with serial communication.

**Comment by pieterh** [comment]
I have the impression that performing the IO with the serial port in an isolated process, is a conscious design decision. For me it isn't clear what the reasoning behind this is. 
Anyways, you should be able to start the SharpSerial.dll by executing the 'dotnet' as the filename and pass the string with the assembly location as arguments. Of course concatenate your own arguments to start the SharpSerial program.
Do not change the extension. For net6.0 the output is a .dll and this should be fine.
Since you probably also need to support 4.0, you could determine the extension of the assembly. In case it is an exe, just start the proces as you did in the past. If the extension is a .dll, use 'dotnet' to start the process.
Unfortunately I'm not able to run the unit test on my mac properly. However, I do see that de .dll is started with the proper arguments and some communication with this child process is going on.

**Comment by samuelventura** [comment]
I build PC controlled automatic testing machines and sometimes you have to deal with serial devices with buggy windows drivers that instead of failing or throwing an exception crash the whole process. The solution to reach the customer expected reliability was to move serial handling to a separated process tethered back to the main application process through stdio. This is the same approach Erlang ports use to achieve 6 sigma figures. It has worked great for us. 

I was aware of the dotnet option but feels like a smell and may not work in runtime only installs (is dotnet included only in the SDK?). I think the cleaner approach is to split this library or reverse the dependencies which would not be noticeable for those requiring both modbus/tcp and serial/rtu. My Elixir modbus library takes this approach. The serial RTU part depends on the TCP part to make use of the common behaviours (interfaces). An even cleaner approach would be to have it further split into core, tcp, and rtu/serial. There could also be a net6.0 branch with it own package that drops the serial part. 

This repo is not being use to support any of our products any more. The new guy in charge of those products forked this some time ago. This repo is here for historical reason and may be archived soon mostly because I am all Erlang/Elixir/Nerves these days.

**Comment by pieterh** [comment]
hi, thank you for the background information. Now it makes sense to me. I agree with you that reversing the dependencies would be the best option. and splitting up even better. With the isolated IO stream you could run into more problems in the future with platform differences. With the core, tcp (and normal rtu stream) less likely.
Do you prefer to do it yourself or would you like me to give it a go? If i set it up, I definitely need you to test the RTU package.

**Comment by samuelventura** [comment]
I would be great to count on your help to upgrade this. Go ahead. Thank you.

**Comment by pieterh** [comment]
Before starting to make changes, I would like to check the approach with you first. I assume that if someone will use it with an RTU, then it is likely that the isolated process is needed. 

SharpModbus:
- move the RTU related stuf to the new project/nuget package 'SharpModusRTU' (see below)
- drop dependency on SharpSerial
- is now clean of dependencies
- will only contain the TCP stream & protocol
- should be able to run fine on net40/net6.0/net7.0
- breaking changes -> version 2?

SharpSerial:
- don't include manual .exe
- revert the changes in main for net60?
- needs a new version 1.0.8 which would be 
- should be able to run fine on net40
- if in the future is a need to get it to a higher .net version, the isolated process stuff needs to be fixed

SharpModusRTU: (the new project and package)
- not sure about the name ;-)
- depends on SharpSerial and SharpModbus
- should be able to run fine on net40

What do you think of the above?

**Comment by samuelventura** [comment]
I would be in favor of keeping pure code (including RTU) in SharpModbus and moving all Isolated stuff to SharpSerial and reverse the dependecy. I just noticed/remembered that SharpSerial also has some Property Grid editor stuff that should stay there. SerialSettings is currently a DTO and would need an interface to be moved to SharpModbus. 

This is what I think would be the minimal-change new layout:

SharpModbus: All modbus (both tcp/rtu) suff and standard serial stuff (including serialsettings interface).
SharpSerial: Isolated serial stuff and property grid editor stuff (Windows only).

In this ways we future proof SharpModbus and move all the specialiced and hard to maintain stuff to SharpSerial.

The layout above seems to achieve the same advantages you got (clean of dependencies, run fine on net40/net6.0/net7.0), and maybe even become cross platform, while keeping both tcp and rtu protocols on the same package.

SharpSerial on the other hand may be left net40 only until someone asks for the upgrade since I suspect no body is using those specialiced features.

**Comment by pieterh** [comment]
ok, i just refactored it, as you recommended, in my fork. The projects build fine and SharpModbus tests is running properly. The SharpSerial tests are not working on my mac. I will see if I can test it tomorrow on a windows system. I build a test nuget package for the SharpModbus and used it as a local source for my project. That works fine without modifications and build warnings.
Tomorrow I will double check my changes to see if it looks nice. Will give you an update after that.

**Comment by pieterh** [comment]
finished the refactoring of both packages and  for as I can tell it works fine. Unfortunately I was not able to test with a real serial interface. I have created for both packages an pull request. Could you have a look at it?

**Comment by samuelventura** [comment]
Just added a Socat subproject to test serial on MacOS/Linux. It is green in my kubuntu and macpro. Could you check on your side before publishing? No extra hardware required. Great work. Thanks.

**Comment by samuelventura** [comment]
Just added a Com0com subproject to test serial on Windows. It is green in my Win10x64 VM.

**Comment by pieterh** [comment]
I have tested it and it looks fine to me. After installing socal, that test project works also good. Maybe you want to have a look at the namespace of these two projects? They contain the text 'special' in them and this doesn't match the project name. Not sure if it was intended, but I would like to mention that it might be confusing.

**Comment by samuelventura** [comment]
Good catch.  I just rename those projects and namespaces after the target OS to make them the place for OS specific test from now on. Let me know if anything else strikes you as odd.

**Comment by pieterh** [comment]
It looks nice to me!

**Comment by samuelventura** [comment]
Just published the [v2.0.0 release](https://github.com/samuelventura/SharpModbus/releases/tag/v2.0.0) and [nuget package.](https://www.nuget.org/packages/SharpModbus/2.0.0) Thanks for all your help.

**Comment by pieterh** [comment]
well... thank you for the modbus library! It saved me a lot of work. This is the least I could do to make sure it is up-to-date with the latest .net releases.
